### DISCUSSION TITLE
Getting ArgumentOutOfRangeException on line with no collection, argument or index!

### DISCUSSION THREAD

**Discussion by SoftCircuits** [dotnet/csharplang#8419] [General]
I have a Razor Pages web application (.NET 8.0) with the following code in my markup.

    @for (int i = 0; i < rows; i++)
    {
        if (i < Model.DepartTruck.ProductSources.Count)
            isStorage = Model.DepartTruck.ProductSources[i].IsStorage;
        else
            isStorage = Model.IsStorage;

        // Markup

    }

It's possible for `Model.DepartTruck.ProductSources` to be empty. That is the reason for this `if` statement. However, when I run it, an `ArgumentOutOfRangeException` is thrown.

![image](https://github.com/user-attachments/assets/bcde4297-aba6-471a-892e-a3e92c2d8543)

In this case, `Model.DepartTruck.ProductSources` is empty and `i` is zero. So why the exception? And look at the line where the exception occurs! The highlighted line has no collection, no argument and no index.

`Model` is my page model. And `IsStorage` is a public `bool` variable in the model.

Does anyone know why I am seeing this?

**Comment by CyrusNajmabadi** [comment]
Not a question about the design of the c# language.  You could try stack overflow.  Or one of the discord channels linked to from dotnet/Roslyn.
